lib/libuzfs/libuzfs.c

```diff
@@ -1239,40 +1360,45 @@ libuzfs_object_write(libuzfs_dataset_handle_t *dhp, uint64_t obj,
 		buf += nwrite;
 	}
 
-	if (sync) {
+	zfs_rangelock_exit(lr);
+	libuzfs_return_node(dhp, up);
+
+	if (sync && err == 0) {
 		zil_commit(dhp->zilog, obj);
 	}
 
 out:
 	dnode_rele(dn, FTAG);
-	sa_handle_destroy(sa_hdl);
 	return (err);
 }
 
 int
 libuzfs_object_read(libuzfs_dataset_handle_t *dhp, uint64_t obj,
     uint64_t offset, uint64_t size, char *buf)
 {
-	int err = 0;
-	uint64_t obj_size = 0;
-	uint64_t read_size = 0;
-	objset_t *os = dhp->os;
-
-	err = libuzfs_object_get_size(dhp, obj, &obj_size);
-	if (err)
-		return (-err);
+	libuzfs_node_t *up;
+	int rc = libuzfs_request_node(dhp, obj, &up);
+	if (rc != 0) {
+		return (-rc);
+	}
 
-	if (offset > obj_size)
-		return (0);
+	zfs_locked_range_t *lr = zfs_rangelock_enter(&up->rl,
+	    offset, size, RL_READER);
+	if (offset >= up->u_size) {
+		goto out;
+	}
 
-	read_size = offset + size > obj_size ? obj_size - offset : size;
+	rc = MIN(up->u_size - offset, size);
```

Comment:
it's better to use read_size instead of rc for this, "rc" is only used for return value, we could use "rc = -err" or "rc = read_size" to return, but we should not use rc standing for read_size.